### PR TITLE
Implement Logger middleware regex exclude pattern

### LINKED ISSUE #1682: Logger middleware exclude pattern
So that a regex can be used for exclusion, for path prefixes and such.

```rust
impl Logger {
	fn exclude_pattern(mut self, pattern: impl Pattern) -> Self
}
```

### PR DESCRIPTION
**Pull request by mattgathu**
## PR Type

Feature

## PR Checklist
Check your PR fulfills the following:

- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] A changelog entry has been made for the appropriate packages.
- [x] Format code with the latest stable rustfmt

## Overview
Support use of regex for logger exclusion rules

part of #1682

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -9,13 +9,15 @@
   to retain any trailing slashes. [#1695]
 * Remove bound `std::marker::Sized` from `web::Data` to support storing `Arc<dyn Trait>`
   via `web::Data::from` [#1710]
+* Implement Logger middleware regex exclude pattern [#1723]
```
Comment (robjtede) [inline CHANGES.md]: move this and link line to unreleased section

File: src/middleware/logger.rs [hunk 2]
```diff
@@ -123,6 +138,7 @@ impl Default for Logger {
         Logger(Rc::new(Inner {
             format: Format::default(),
             exclude: HashSet::new(),
+            exclude_regex: RegexSet::new(Vec::<&str>::new()).unwrap(),
```
Comment (robjtede) [inline src/middleware/logger.rs]:
```suggestion
            exclude_regex: RegexSet::empty(),
```

File: src/middleware/logger.rs [hunk 3]
```diff
@@ -100,6 +101,7 @@ impl Logger {
         Logger(Rc::new(Inner {
             format: Format::new(format),
             exclude: HashSet::new(),
+            exclude_regex: RegexSet::new(Vec::<&str>::new()).unwrap(),
```
Comment (robjtede) [inline src/middleware/logger.rs]: I got `RegexSet::empty` added to `regex` v1.4.

```suggestion
            exclude_regex: RegexSet::empty(),
```

Comment (robjtede) [reply src/middleware/logger.rs]: also will need the cargo toml regex version bumping

File: src/middleware/logger.rs [hunk 4]
```diff
@@ -111,6 +113,19 @@ impl Logger {
...
         self
     }
+
+    /// Ignore and do not log access info for paths that match regex
+    pub fn exclude_regex<T: Into<String>>(
+        mut self,
+        path: T,
+    ) -> Result<Self, regex::Error> {
+        let inner = Rc::get_mut(&mut self.0).unwrap();
+        let mut patterns: Vec<String> = inner.exclude_regex.patterns().into();
```
Comment (robjtede) [inline src/middleware/logger.rs]:
```suggestion
        let mut patterns = inner.exclude_regex.patterns().to_vec();
```

**Review by robjtede — CHANGES_REQUESTED** [review]
Looks alright so far. I think that middleware constructors should panic on invalid input so we can either just panic or require that a regex is passed in that we extract the known good pattern from.

**Comment by mattgathu** [comment]
> Looks alright so far. I think that middleware constructors should panic on invalid input so we can either just panic or require that a regex is passed in that we extract the known good pattern from.

I went with panic, to avoid the user having to rely on the regex crate and also avoid incompatibility if the user's regex crate version significantly differs from the one used by Actix.

**Review by robjtede — APPROVED** [review]
I agree that not forwarding (leaking) the regex error type is a compelling reason to panic instead. Thanks for tackling this.

**Pull request merged**